daemon/graphdriver/projectquota.go

```diff
@@ -0,0 +1,292 @@
+// +build linux
+
+//
+// projectquota.go - implements XFS project quota controls
+// for setting quota limits on a newly created directory.
+// It currently supports the legacy XFS specific ioctls.
+//
+// TODO: use generic quota control ioctl FS_IOC_FS{GET,SET}XATTR
+//       for both xfs/ext4 for kernel version >= v4.5
+//
+
+package graphdriver
+
+/*
+#include <stdlib.h>
+#include <dirent.h>
+#include <linux/fs.h>
+#include <linux/quota.h>
+#include <linux/dqblk_xfs.h>
+struct fsxattr {
+	__u32		fsx_xflags;
+	__u32		fsx_extsize;
+	__u32		fsx_nextents;
+	__u32		fsx_projid;
+	unsigned char	fsx_pad[12];
+};
+#define FS_XFLAG_PROJINHERIT	0x00000200
+#define FS_IOC_FSGETXATTR		_IOR ('X', 31, struct fsxattr)
+#define FS_IOC_FSSETXATTR		_IOW ('X', 32, struct fsxattr)
+
+#define PRJQUOTA	2
+#define XFS_PROJ_QUOTA	2
+#define Q_XSETPQLIM QCMD(Q_XSETQLIM, PRJQUOTA)
+#define Q_XGETPQUOTA QCMD(Q_XGETQUOTA, PRJQUOTA)
+*/
+import "C"
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"syscall"
+	"unsafe"
+
+	"github.com/Sirupsen/logrus"
+)
+
+// Quota limit params - currently we only control blocks hard limit
+type Quota struct {
+	Size uint64
+}
+
+// QuotaCtl - Context to be used by storage driver (e.g. overlay)
+// who wants to apply project quotas to container dirs
+type QuotaCtl struct {
+	backingFsDev  string
+	nextProjectID uint32
+	quotas        map[string]uint32
+}
+
+// NewQuotaCtl - initialize project quota support by finding the first project id
+// to be used for the next container create
+func NewQuotaCtl(basePath string) (*QuotaCtl, error) {
+	//
+	// get max project id for next use
+	//
+	maxProjectID, err := getMaxProjectID(basePath)
```

Comment:
We do not get to directory scanning, because the first thing that getMaxProjectID() does is get the project id of the home directory. This test will fail if no pquota support.
If pquota is supported and home directory does not have a project id the check will succeed and return value of 0.
There is also a 'hidden' feature here - project quota can be used to set a quota limit (using xfs_quota tool) on the entire storage driver directory (/var/lib/docker/overlay2), to set it apart from the rest of the system. In that case, the root project id will be used as a "start offset" (e.g. echo docker:999 > /etc/projid) and all containers will be assigned project ids > 1000.
This is a way to prevent xfs_quota management from conflicting with docker.
I did not document this feature because I did not want to complicate. It can be documented later on in advanced documentation.